iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -10,568 +10,728 @@
+                                                      reported_state_callback, twin_data_context);
+    }
+
+    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result,
+                     "IoTHub(Device|Module)Client_SendReportedState failed.");
+}
+
+static void _receive_reported_state_loop(REPORTED_TWIN_DATA* reported_twin_data)
+{
+    time_t begin_operation = time(NULL);
+    time_t now_time;
+
+    LogInfo("_receive_reported_state_loop(): Entering loop.\n");
+
+    while (now_time = time(NULL), difftime(now_time, begin_operation) < MAX_CLOUD_TRAVEL_TIME)
+    {
+        // Receive IoT Hub response.
+        if (Lock(reported_twin_data->lock) != LOCK_OK)
         {
-            LogError("Lock_Init failed");
-            free(retValue);
-            retValue = NULL;
+            ASSERT_FAIL("Lock failed.");
         }
         else
         {
-            retValue->receivedCallBack = false;
-            retValue->cb_payload = NULL;
+            if (reported_twin_data->received_callback)
+            {
+                Unlock(reported_twin_data->lock);
+                break;
+            }
+            Unlock(reported_twin_data->lock);
         }
+        ThreadAPI_Sleep(1000);
     }
-    return retValue;
-}
+    ASSERT_IS_TRUE(difftime(now_time, begin_operation) < MAX_CLOUD_TRAVEL_TIME,
+                   "Timeout waiting for twin message.");
 
-static void device_desired_deinit(DEVICE_DESIRED_DATA *device)
-{
-    if (device == NULL)
-    {
-        LogError("invalid parameter device");
-    }
-    else
-    {
-        free(device->cb_payload);
-        Lock_Deinit(device->lock);
-        free(device);
-    }
+    LogInfo("_receive_reported_state_loop(): Exiting loop.\n");
 }
 
-static void device_desired_reset(DEVICE_DESIRED_DATA *device)
+static void _send_event_async(IOTHUB_MESSAGE_HANDLE msgHandle)
 {
-    if (device == NULL)
-    {
-        LogError("invalid parameter device");
-    }
-    else if (Lock(device->lock) == LOCK_ERROR)
+    IOTHUB_CLIENT_RESULT result;
+
+    if (iothub_moduleclient_handle != NULL)
     {
-        LogError("Lock failed");
+        result = IoTHubModuleClient_SendEventAsync(iothub_moduleclient_handle, msgHandle, NULL,
+                                                   NULL);
     }
     else
     {
-        device->receivedCallBack = false;
-        (void) Unlock(device->lock);
+        result = IoTHubDeviceClient_SendEventAsync(iothub_deviceclient_handle, msgHandle, NULL,
+                                                   NULL);
     }
+
+    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result,
+                     "IoTHub(Device|Module)Client_SendEventAsync failed.");
 }
 
-void client_create_with_properies_and_send_d2c(IOTHUB_PROVISIONED_DEVICE* deviceToUse, MAP_HANDLE mapHandle)
-{
-    IOTHUB_MESSAGE_HANDLE msgHandle;
 
+void _client_create_with_properties_and_send_d2c(IOTHUB_PROVISIONED_DEVICE* device_to_use,
```

Comment:
Nit: Since `_client_create_with_properties_and_send_d2c` is only ever used to send the KillTCP message, could we rename thos to like `send_tcp_kill` or something and then move all the code around setting up the MAP_HANDLE in functions below into this one?

I think this would also benefit from some level of comments since this isn't documented as far as I know.  So something like "// IoT Hubs can be configured so that a special device telemetry message will cause them to close the TCP connection.  // This is used to simulate network failures and check the SDK's reconnection logic"